File: code/dart/lib/src/graphql/runtime/graphql_runtime_options.dart
import 'dart:async';

import 'package:modular_api/src/graphql/catalog/graphql_catalog_builder.dart';
import 'package:modular_api/src/graphql/read/sql_read_contract.dart';
import 'package:modular_api/src/graphql/schema/graphql_schema_sdl_generator.dart';

typedef GraphqlCatalogFactory = FutureOr<GraphqlCatalog> Function();
typedef GraphqlSdlFactory = String Function(GraphqlCatalog catalog);
typedef GraphqlEventSink = FutureOr<void> Function(GraphqlRequestEvent event);

const graphqlDefaultSqlReadExecutorCapabilityId = 'modular_api.sql.read_executor';

enum GraphqlRequestPhase {
  started,
  completed,
}

final class GraphqlRequestEvent {
  const GraphqlRequestEvent({
    required this.phase,
    required this.requestId,
    required this.method,
    required this.path,
    this.statusCode,
  });

  final GraphqlRequestPhase phase;
  final String requestId;
  final String method;
  final String path;
  final int? statusCode;
}

final class GraphqlOptions {
  GraphqlOptions({
    required this.catalogFactory,
    this.executor,
    this.executionCapabilityId,
    this.introspectionEnabled = false,
    this.maxDepth = 8,
    this.maxComplexity = 500,
    this.defaultLimit = 50,
    this.maxLimit = 200,
    this.onEvent,
    GraphqlSdlFactory? sdlFactory,
  }) : sdlFactory = sdlFactory ?? const GraphqlSchemaSdlGenerator().generate {
    if (executor != null && executionCapabilityId != null) {
      throw ArgumentError(
        'GraphQL runtime accepts either a direct executor or an execution capability id, not both.',
      );
    }
  }

  final GraphqlCatalogFactory catalogFactory;
  final SqlReadExecutor? executor;
  final String? executionCapabilityId;
  final bool introspectionEnabled;
  final int maxDepth;
  final int maxComplexity;
  final int defaultLimit;
  final int maxLimit;
  final GraphqlEventSink? onEvent;
  final GraphqlSdlFactory sdlFactory;

  String get resolvedExecutionCapabilityId =>
      executionCapabilityId ?? graphqlDefaultSqlReadExecutorCapabilityId;
}

File: code/ts/src/graphql/runtime/graphql_runtime_options.ts
import type { GraphqlCatalog } from '../catalog/graphql_catalog_builder';
import type { ReadExecutor } from '../read/sql_read_contract';
import { GraphqlSchemaSdlGenerator } from '../schema/graphql_schema_sdl_generator';

export type GraphqlEventSink = (event: GraphqlRequestEvent) => void | Promise<void>;
export type GraphqlSourceDigestFactory = () => string | Promise<string>;

export enum GraphqlRequestPhase {
  Started = 'started',
  Completed = 'completed',
}

export class GraphqlRequestEvent {
  readonly phase: GraphqlRequestPhase;
  readonly requestId: string;
  readonly method: string;
  readonly path: string;
  readonly statusCode?: number;

  constructor(options: {
    phase: GraphqlRequestPhase;
    requestId: string;
    method: string;
    path: string;
    statusCode?: number;
  }) {
    this.phase = options.phase;
    this.requestId = options.requestId;
    this.method = options.method;
    this.path = options.path;
    this.statusCode = options.statusCode;
  }
}

export const graphqlDefaultReadExecutorCapabilityId = 'modular_api.sql.read_executor';

export class GraphqlOptions {
  readonly catalogFactory: () => Promise<GraphqlCatalog>;
  readonly executor?: ReadExecutor;
  readonly executionCapabilityId?: string;
  readonly introspectionEnabled: boolean;
  readonly maxDepth: number;
  readonly maxComplexity: number;
  readonly defaultLimit: number;
  readonly maxLimit: number;
  readonly onEvent?: GraphqlEventSink;
  readonly artifactDirectory?: string;
  readonly sourceDigestFactory?: GraphqlSourceDigestFactory;
  readonly sdlFactory: (catalog: GraphqlCatalog) => string;

  constructor(options: {
    catalogFactory: () => Promise<GraphqlCatalog>;
    executor?: ReadExecutor;
    executionCapabilityId?: string;
    introspectionEnabled?: boolean;
    maxDepth?: number;
    maxComplexity?: number;
    defaultLimit?: number;
    maxLimit?: number;
    onEvent?: GraphqlEventSink;
    artifactDirectory?: string;
    sourceDigestFactory?: GraphqlSourceDigestFactory;
    sdlFactory?: (catalog: GraphqlCatalog) => string;
  }) {
    if (options.executor && options.executionCapabilityId) {
      throw new Error(
        'GraphQL runtime accepts either a direct executor or an execution capability id, not both.',
      );
    }

    this.catalogFactory = options.catalogFactory;
    this.executor = options.executor;
    this.executionCapabilityId = options.executionCapabilityId;
    this.introspectionEnabled = options.introspectionEnabled ?? false;
    this.maxDepth = options.maxDepth ?? 8;
    this.maxComplexity = options.maxComplexity ?? 500;
    this.defaultLimit = options.defaultLimit ?? 50;
    this.maxLimit = options.maxLimit ?? 200;
    this.onEvent = options.onEvent;
    this.artifactDirectory = options.artifactDirectory;
    this.sourceDigestFactory = options.sourceDigestFactory;
    this.sdlFactory = options.sdlFactory ?? ((catalog) => new GraphqlSchemaSdlGenerator().generate(catalog));
  }

  get resolvedExecutionCapabilityId(): string {
    return this.executionCapabilityId ?? graphqlDefaultReadExecutorCapabilityId;
  }
}